File: Python/lab21/Order.py
from CardPayment import CardPayment
from CashPayment import CashPayment
from CheckPayment import CheckPayment


class Order:
    def __init__(self, name, type):
        self.name = name
        self.type = type
        self.payment_method = None
        self.is_closed = False

    def set_payment_method(self, payment_method):
        if not self.is_closed:
            if isinstance(payment_method, CardPayment):
                print("Type of payment was changed to CardPayment")
            elif isinstance(payment_method, CashPayment):
                print("Type of payment was changed to CashPayment")
            elif isinstance(payment_method, CheckPayment):
                print("Type of payment was changed to CheckPayment")

            self.payment_method = payment_method
        else:
            print("The order is paid")

    def do_payment(self):
        if not self.is_closed:
            self.payment_method.do_payment()
            self.is_closed = True
        else:
            print(f"The order: {self.name} type: {self.type} is paid")

File: Java/lab21/src/main/java/org/example/Order.java
package org.example;

public class Order {
    private final String name;
    private final String type;
    private PaymentMethod paymentMethod;
    boolean isClose;

    public Order(String name, String type) {
        this.name = name;
        this.type = type;
        isClose=false;
    }
    public void setPaymentMethod(PaymentMethod paymentMethod) {
        if(!isClose){
            if (paymentMethod instanceof CardPayment){
                System.out.println("Type of payment was changed to CardPayment");
            }else if (paymentMethod instanceof CashPayment){
                System.out.println("Type of payment was changed to CashPayment");
            }else if (paymentMethod instanceof CheckPayment){
                System.out.println("Type of payment was changed to CheckPayment");
            }
            this.paymentMethod = paymentMethod;
        }else {
            System.out.println("The order is paid");
        }

    }

    public void doPayment(){
        if(!isClose){
            paymentMethod.doPayment();
            isClose=true;
        }else {
            System.out.println("The order:"+name+" type: "+type+" is paid");
        }
    }
}
